sympy/functions/special/bessel.py

```diff
@@ -216,11 +217,30 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
     def _eval_rewrite_as_jn(self, nu, z, **kwargs):
         return sqrt(2*z/pi)*jn(nu - S.Half, self.argument)
 
+    def _eval_as_leading_term(self, x, cdir=0):
+        nu, z = self.args
+        arg = z.as_leading_term(x)
+        if x in arg.free_symbols:
+            return z**nu
+        else:
+            return self.func(*self.args)
+
     def _eval_is_extended_real(self):
         nu, z = self.args
         if nu.is_integer and z.is_extended_real:
             return True
 
+    def _eval_nseries(self, x, n, logx, cdir=0):
+        from sympy.series.order import Order
+        nu, z = self.args
+
+        if z.limit(x, 0) is S.Zero:
+            s = [(-1)**k * z**(2 * k) / (factorial(k) * factorial(nu + k)
+                    * 2**(2 * k)) for k in range((n + 1)//2)] + [Order(z**(n))]
```

Comment:
More terms are needed if the order of z in x is less than 1, eg.  sqrt(x), cbrt(x)... (The range can be computed from the leading term of z.) I would also add the last Order term only at the return line so it would be `Order(x**n)` which may be different from `Order(z**n)`.

Reply:
Currently - 
```
In [8]: besselj(0, sqrt(x)).series(x)
Out[8]: 
         2     3       4         5           
    x   x     x       x         x        ⎛ 6⎞
1 - ─ + ── - ──── + ────── - ──────── + O⎝x ⎠
    4   64   2304   147456   14745600        

In [9]: besselj(1, sqrt(x)).series(x)
Out[9]: 
      3/2    5/2     7/2      9/2       11/2          
√x   x      x       x        x         x          ⎛ 6⎞
── - ──── + ──── - ───── + ─────── - ───────── + O⎝x ⎠
2     16    384    18432   1474560   176947200        

```
It seems like there are sufficient terms and I think the caller is taking care of n?

Reply:
There are sufficiently many terms because `_eval_nseries` is called repeatedly from `expr.series` with increasing values of `n`:  https://github.com/sympy/sympy/blob/b564d9ba2705ee2978766e1ee2102750834df68b/sympy/core/expr.py#L3002-L3016
However, that is inefficient and we would want to have a sufficient number of terms on the first call. (This whole piece of code looks like a hack that was created to mask errors in various `_eval_nseries` methods. Besides, it will add at most 9 terms which is not enough in general.)

Reply:
Oh, I hadn't noticed that. Then perhaps, the block of code you pointed out should be removed in the future, by making appropriate corrections to the `_eval_nseries` methods of functions that need it.

Reply:
Only two tests in the `series` module fail on commenting this block - 
```py
assert D(exp(x), x).series(n=3) == D(1, x) + D(x, x) + D(x**2/2, x) + D(x**3/6, x) + O(x**3) # where D = Derivative
assert Derivative(sin(x), x).series(x, n=3).doit() == cos(x).series(x, n=3)
```
will look into this